### PR TITLE
[BEAM-3194] Fail if @RequiresStableInput is used on runners that don't supoort it

### PR DESCRIPTION
**Pull request by robinyqiu**
r: @lukecwik 
cc: @kennknowles @aaltay

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by lukecwik** [comment]
I see you have 5 runners listed, does that mean Dataflow and DirectRunner/Reference runner already support it?

**Comment by robinyqiu** [comment]
It's supported on Dataflow. DirectRunner doesn't have retry do it doesn't matter. I am not very sure how the Reference runner is implemented in terms of retry. Do you know the details or who should I talk to?

**Comment by lukecwik** [comment]
Talk to @youngoli

File: runners/spark/src/main/java/org/apache/beam/runners/spark/SparkTransformOverrides.java [hunk 1]
```diff
@@ -23,13 +23,20 @@
...
 import org.apache.beam.runners.core.construction.SplittableParDoNaiveBounded;
+import org.apache.beam.runners.core.construction.UnsupportedOverrideFactory;
 import org.apache.beam.sdk.runners.PTransformOverride;
 import org.apache.beam.sdk.transforms.PTransform;
 
 /** {@link PTransform} overrides for Flink runner. */
 public class SparkTransformOverrides {
   public static List<PTransformOverride> getDefaultOverrides(boolean streaming) {
     ImmutableList.Builder<PTransformOverride> builder = ImmutableList.builder();
+    // TODO: [BEAM-5358] Support @RequiresStableInput on Spark runner
```
Comment (twosom) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/SparkTransformOverrides.java:34]: have beam community have plan for support `@RequiresStableInput` annotation for SparkRunner?

Comment (aaltay) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/SparkTransformOverrides.java:34]: I do not believe there is a plan, but feel free to ask on the dev@ email list.

**Pull request merged**